File: Exercise_4/MonteCarlo.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Feb 15 09:25:46 2022

@author: siriapasini
"""
import math as m
import numpy as np
import matplotlib.pyplot as plt
from numba import njit
from random import random
import AppKit

Np = 32 #number of particles
sigma = 0.2556 
n0 = 21.86 
epsilon = 10.22

xpos = np.zeros((Np))
ypos = np.zeros((Np))
zpos = np.zeros((Np))
dx = np.zeros([Np,Np])
dy = np.zeros([Np,Np])
dz = np.zeros([Np,Np])
r = np.zeros([Np,Np])
der = np.zeros([Np,Np])
der2 = np.zeros([Np,Np])


def distance(x,y,z):
    for i in range(0,Np):
        for j in range(0,Np):
            if i!=j:
                dx[i,j] = x[i]-x[j]
                dx[i,j] = dx[i,j] - L* np.rint(dx[i,j]/L)
                dy[i,j] = y[i]-y[j]
                dy[i,j] = dy[i,j] - L* np.rint(dy[i,j]/L)
                dz[i,j] = z[i]-z[j]
                dz[i,j] = dz[i,j] - L* np.rint(dz[i,j]/L)
                r[i,j] = (dx[i,j]**2 +dy[i,j]**2 +dz[i,j]**2)**0.5
                
    return r,dx,dy,dz


def probability(x, y,z, b):
    P = 0
    p = 0
    r,dx,dy,dz = distance(x,y,z)
    #print("{}".format(r))
    for i in range(0,Np):
        for j in range(0,i):
            if i!=j:
                p += (-b/r[i,j])**5

    P = np.exp(p)
    
    return P,r,dx,dy,dz

def sampling(xpos,ypos,zpos, delta, acc,rej, b):
    rx = np.zeros((Np))
    ry = np.zeros((Np))
    rz = np.zeros((Np))
    for i in range (0,Np):
        rx[i] = delta*(random()-0.5)
        ry[i] = delta*(random()-0.5)
        rz[i] = delta*(random()-0.5)
    xprop = xpos+ rx #generate proposed new positions
    xprop = xprop - L * np.rint(xprop/L)
    yprop = ypos+ry
    yprop = yprop - L * np.rint(yprop/L)
    zprop = zpos+rz
    zprop = zprop - L * np.rint(zprop/L)

    
    #evaluate ratio
    ratio = probability(xprop,yprop,zprop,b)[0]/ probability(xpos,ypos,zpos,b)[0]
    #print("{}".format(ratio))
    p = min(ratio,1)
    r2 = random()


    if p>r2:
        xpos = xprop
        ypos = yprop
        zpos = zprop
        acc += 1
    else: 
        rej += 1
    
    P,r,dx,dy,dz = probability(xpos,ypos,zpos,b)
    
    E, V,T, Tjf = energy(xpos, ypos, zpos, Np,b, r,dx,dy,dz)
    
    return xpos,ypos,zpos, acc, rej, E, V, T, Tjf   


def potential(r):
    V = 0
    for i in range(0,Np):
        for j in range(0,i):
            if i!=j:
                V += 4*((1/r[i,j])**12 - (1/r[i,j])**6)            
    return V

def kinetic(xpos,ypos,zpos,Np, b,r,dx,dy,dz):
    T =0
    Gx = np.zeros(Np)
    Gy = np.zeros(Np)
    Gz = np.zeros(Np)
    t = np.zeros(Np)    
    tjf = np.zeros(Np)
    Tjf=0

    for l in range (0,Np):
        for i in range (0,Np):
            if i!=l:                
                t[l] += 10*b**5 / (r[l,i]**7) 
                tjf[l] += 5*b**5 / (r[l,i]**7) 
        
     
            for k in range (0,Np):
                if k!=l and  i!=l:
                    Gx[l] += -(25/4) * b**10 * dx[l,i]*dx[l,k] / (r[l,i]**7 * r[l,k]**7)
                    Gy[l] += -(25/4) * b**10 * dy[l,i]*dy[l,k] / (r[l,i]**7 * r[l,k]**7)
                    Gz[l] += -(25/4) * b**10 * dz[l,i]*dz[l,k] / (r[l,i]**7 * r[l,k]**7)
        
        
        T += t[l] + Gx[l]+ Gy[l]+Gz[l] 
        Tjf += tjf[l]
        
    T = T * 0.09094
    Tjf = Tjf * 0.09094    
    
    return  T, Tjf


def energy(xpos,ypos,zpos,Np, b,r,dx,dy,dz):
    V = potential(r)
    T, Tjf = kinetic(xpos,ypos,zpos,Np,b,r,dx,dy,dz)
    E = V + T
    
    return E, V, T, Tjf

# INITIAL POSITION IN THE LATTICE (FOR NOW THIS IS OK, I USED THE ONE DEGIO DID, BUT I HAVE TO CHECK THIS AGAIN)
a = 0.5 * (Np / n0)**(1.0/3.0) /sigma
L = 2*a


n = 0

for k in range(0,4):
    for i in range(0,4):
        for j in range(0,2):
            zpos[n] = k * a / 2 # set z coordinate
            xpos[n] = i * a / 2 # set x coordinate
            if k%2 == 0:
        
                ypos[n] = j * a + i%2 * a/2 # set y coordinate
        
            else:
                
                ypos[n] = (j + 1/2) * a - i%2 * a/2 # set y coordinate
            

            n= n+1
            


fig = plt.figure()
ax = fig.add_subplot(projection='3d')
ax.scatter(xpos,ypos,zpos, marker=".")
ax.set_xlabel("x")
ax.set_ylabel("y")
ax.set_zlabel("z")
plt.title("Initial Conditions")



# EQUILIBRATION STEP AND DETERMINATION OF DELTA
Ne = 1000 #equilibration number of steps
eff = 1
delta = 0.11 #initial choice just to have a starting point
b = 1.2

# Equilibration step
print("Starting equilibration and definition of delta")
while abs (eff - 0.5) > 0.1:
    acc = 0 #acceptance number
    rej = 0 #rejection number
    Esum = 0
    Esum2 = 0
    for j in range (0,Ne):
        xpos,ypos,zpos, acc, rej, E, V, T, Tjf  = sampling(xpos,ypos,zpos, delta,acc,rej,b)
        N = acc + rej
        eff = acc / N
        print("{}".format(j))
    delta = delta + 0.01

delta = delta - 0.01 
print("I found delta to be equal to {}".format(delta))  


fig = plt.figure()
ax = fig.add_subplot(projection='3d')
ax.scatter(xpos,ypos,zpos, marker=".")
ax.set_xlabel("x")
ax.set_ylabel("y")
ax.set_zlabel("z")
plt.title("Actual Initial Conditions")


Nsteps = 10000

Nsamples = 1
I = np.zeros((Nsamples))
sigma = np.zeros((Nsamples))
b = np.zeros((Nsamples)) #variational parameter
b[0] = 1.2


for i in range (0,Nsamples):
    print("Value of b = {}".format(b[i]))
    acc = 0 #acceptance number
    rej = 0   
    Esum = 0
    Esum2 = 0
    E = np.zeros(Nsteps)
    V = np.zeros(Nsteps)
    T = np.zeros(Nsteps)
    Tjf = np.zeros(Nsteps)
    
    
    for j in range (0,Nsteps):
        xpos,ypos,zpos, acc, rej, E[j], V[j], T[j], Tjf[j]  = sampling(xpos,ypos,zpos, delta,acc,rej,b[i])
        N = acc + rej
        eff = acc / N
        Esum += E[j]
        Esum2 += E[j]**2
         
        print("{}".format(j))
        
    I[i] = Esum/Nsteps
    sigma[i] = (1/(Nsteps-1) * (Esum2/Nsteps - I**2) )**0.5
    print("{}".format(I[i]))

    
    if i != (Nsamples-1):
        b[i+1] = b[i]+ 0.1



fig = plt.figure()
plt.plot(np.linspace(0,Nsteps,Nsteps) , V, linewidth=0.9, label="V_{lj]")
plt.plot(np.linspace(0,Nsteps,Nsteps) , T, linewidth=0.9, label="T_{local}")
#plt.plot(np.linspace(0,Nsteps,Nsteps) , E, linewidth=0.9, label="E_{L}")
#plt.plot(np.linspace(0,Nsteps,Nsteps) , Tjf, linewidth=0.9, label="T_{JF,local}")
plt.legend()
plt.title("") 
plt.savefig("WF_20_K_el.png", dpi=300)
                
        
# # fig = plt.figure()
# # ax = fig.add_subplot(projection='3d')
# # ax.scatter(xpos,ypos,zpos, marker=".")
# # ax.set_xlabel("x")
# # ax.set_ylabel("y")
# # ax.set_zlabel("z")
# # plt.title("Final Conditions")   

# # AppKit.NSBeep()

File: Exercise_4/montecarlo.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>
#include<string.h>
#include<time.h>


int Np = 32; //number of particles
double sigma = 0.2556; //unit of length
double n0 = 21.86; //saturation density
double epsilon = 10.22; // unit of energy
double b = 1.23;

int main(void)
{
  double dr[Np][Np];
  double dx[Np][Np];
  double dy[Np][Np];
  double dz[Np][Np];
  double r[3][Np];
  double Ne = 10000;
  double delta = 0.11; //from some trials it seems the best value
  double l = 0.5 * cbrt(Np/n0) /sigma;
  double L = 2*l;
  //FILE *fp;
  //fp = fopen("potential.csv","w");
  double rprop[3][Np];
  double drprop[Np][Np];
  double dxprop[Np][Np];
  double dyprop[Np][Np];
  double dzprop[Np][Np];
  double eff;
  srand(time(0));// NEEDED TO GENERATE ALWAYS DIFFERENT RANDOM NUMBERS

// THERMALIZATION AND LOOKING FOR THE CORRECT DELTA TO GET A 50% PROBABILITY
do{
  // Initial position of each particle
  int n = 0;
  for (int k = 0; k<4; k++){
      for (int i = 0; i<4; i++){
          for (int j = 0; j<2; j++){
              r[2][n] = k * l / 2;  // set z coordinate
              r[0][n] = i * l / 2; // set x coordinate

              if (k%2 == 0){
                  r[1][n] = j * l + i%2 * l/2; // set y coordinate
              }
              else {
                  r[1][n] = (j + 1/2) * l - i%2 * l/2; // set y coordinate
              }
              n = n+1;
          }
      }
  }
  int acc = 0;
  int rej = 0;
  printf("%f\n", delta);

  for (int n=0; n<Ne; n++){
      printf("Equilibrium %d\n", n);
      // new positions proposal with attention to boundary conditions
    for (int i=0;i<3;i++){
      for (int j=0;j<Np;j++){
        double r1 = (double)rand()/RAND_MAX;
        //printf("%f\n", r1);
        rprop[i][j] = r[i][j] + delta*(r1-0.5);
        rprop[i][j] = rprop[i][j] - L * round(rprop[i][j]/L);
      }
    }
      // Calculation of all the possible distances in all directions for the 2 sets of positions
    for (int i=0;i<Np;i++){
      for (int j=0;j<Np;j++){
        dx[i][j] = r[0][i]-r[0][j];
        dx[i][j] = dx[i][j] - L* round(dx[i][j]/L);
        dy[i][j] = r[1][i]-r[1][j];
        dy[i][j] = dy[i][j] - L* round(dy[i][j]/L);
        dz[i][j] = r[2][i]-r[2][j];
        dz[i][j] = dz[i][j] - L* round(dz[i][j]/L);

        dr[i][j] = sqrt( pow(dx[i][j],2)+pow(dy[i][j],2)+pow(dz[i][j],2));

        dxprop[i][j] = rprop[0][i]-rprop[0][j];
        dxprop[i][j] = dxprop[i][j] - L* round(dxprop[i][j]/L);
        dyprop[i][j] = rprop[1][i]-rprop[1][j];
        dyprop[i][j] = dyprop[i][j] - L* round(dyprop[i][j]/L);
        dzprop[i][j] = rprop[2][i]-rprop[2][j];
        dzprop[i][j] = dzprop[i][j] - L* round(dzprop[i][j]/L);

        drprop[i][j] = sqrt( pow(dxprop[i][j],2)+pow(dyprop[i][j],2)+pow(dzprop[i][j],2));
        }
      }

      // Calculate probability
    double p = 0.0;
    double pprop = 0.0;
    for(int i=0; i<Np; i++){
        for(int j=0; j<i; j++){
            p += -pow((b/dr[i][j]), 5);
            pprop += -pow((b/drprop[i][j]), 5);
        }
    }


    double ratio = pow(M_E,pprop)/pow(M_E,p);
    //srand(time(0));
    double r2;
    for (int i=0;i<1;i++){
      r2 = (double)rand()/RAND_MAX;
    }
    //printf("%f, %f\n", p, pprop);

    if (ratio>r2){
      acc = acc+1;
      for (int i=0;i<Np;i++){
        for (int j=0;j<Np;j++){
          dr[i][j] = drprop[i][j];
          dx[i][j] = dxprop[i][j];
          dy[i][j] = dyprop[i][j];
          dz[i][j] = dzprop[i][j];
        }
      }
      for (int k=0;k<3;k++){
        for (int i=0;i<Np;i++){
          r[k][i] = rprop[k][i];
        }
      }
    }
    else{
      rej = rej +1;
    }
    //printf("%d\n", acc);
    //printf("%d\n", rej);
  }
  double Nsamples = acc + rej;
  //printf("%f\n", Nsamples);
  eff = acc / Nsamples;
  printf("%f\n", eff);
  //printf("%f\n",delta);
  delta = delta + 0.01;
}while (eff >0.6 || eff<0.4);


delta = delta -0.01;

// ACTUAL CALCULATION OF THE ENERGIES AND SAMPLING PROCUDURE
int acc = 0;
int rej = 0;
int Nsteps = 20000;
double V[Nsteps];
double T[Nsteps];
double E[Nsteps];
double Tjf[Nsteps];

double Esum = 0;
double Esum2 = 0;


for (int n=0; n<Nsteps; n++){
    printf("%d\n", n);
  // new positions proposal with attention to boundary conditions
  for (int i=0;i<3;i++){
    for (int j=0;j<Np;j++){
      double r1 = (double)rand()/RAND_MAX;
      //printf("%f\n", r1);
      rprop[i][j] = r[i][j] + delta*(r1-0.5);
      rprop[i][j] = rprop[i][j] - L * round(rprop[i][j]/L);
    }
  }
    // Calculation of all the possible distances in all directions for the 2 sets of positions
  for (int i=0;i<Np;i++){
    for (int j=0;j<Np;j++){
      dx[i][j] = r[0][i]-r[0][j];
      dx[i][j] = dx[i][j] - L* round(dx[i][j]/L);
      dy[i][j] = r[1][i]-r[1][j];
      dy[i][j] = dy[i][j] - L* round(dy[i][j]/L);
      dz[i][j] = r[2][i]-r[2][j];
      dz[i][j] = dz[i][j] - L* round(dz[i][j]/L);

      dr[i][j] = sqrt( pow(dx[i][j],2)+pow(dy[i][j],2)+pow(dz[i][j],2));

      dxprop[i][j] = rprop[0][i]-rprop[0][j];
      dxprop[i][j] = dxprop[i][j] - L* round(dxprop[i][j]/L);
      dyprop[i][j] = rprop[1][i]-rprop[1][j];
      dyprop[i][j] = dyprop[i][j] - L* round(dyprop[i][j]/L);
      dzprop[i][j] = rprop[2][i]-rprop[2][j];
      dzprop[i][j] = dzprop[i][j] - L* round(dzprop[i][j]/L);

      drprop[i][j] = sqrt( pow(dxprop[i][j],2)+pow(dyprop[i][j],2)+pow(dzprop[i][j],2));
      }
    }

    // Calculate probability
  double p = 0.0;
  double pprop = 0.0;
  for(int i=0; i<Np; i++){
      for(int j=0; j<i; j++){
          p += -pow((b/dr[i][j]), 5);
          pprop += -pow((b/drprop[i][j]), 5);
      }
  }


  double ratio = pow(M_E,pprop)/pow(M_E,p);
  //srand(time(0));
  double r2;
  for (int i=0;i<1;i++){
    r2 = (double)rand()/RAND_MAX;
  }
  //printf("%f, %f\n", p, pprop);

  if (ratio>r2){
    acc = acc+1;
    for (int i=0;i<Np;i++){
      for (int j=0;j<Np;j++){
        dr[i][j] = drprop[i][j];
        dx[i][j] = dxprop[i][j];
        dy[i][j] = dyprop[i][j];
        dz[i][j] = dzprop[i][j];
      }
    }
    for (int k=0;k<3;k++){
      for (int i=0;i<Np;i++){
        r[k][i] = rprop[k][i];
      }
    }
  }
  else{
    rej = rej +1;
  }

// Given the new configuration we calculate the value of the potential
V[n] = 0;
for (int i=0;i<Np;i++){
  for(int j=0; j<i;j++){
    V[n] += 4*(pow(1/dr[i][j],12)- pow(1/dr[i][j],6));
  }
}

// Given the new configuration we calculate the local kinteic energy
T[n] = 0;
Tjf[n] = 0;
double G = 0;

for (int l=0;l<Np;l++){
  for(int i=0; i<Np;i++){
    if (i!=l){
      T[n] += 10*pow(b,5)* pow(1/dr[l][i] , 7);

      for(int k=0; k<Np;k++){
        if(k!=l){
        G += -(25/4) * pow(b,10) * dx[l][i]*dx[l][k] / ( pow(dr[l][i] , 7) * pow(dr[l][k] , 7));
        G += -(25/4) * pow(b,10) * dy[l][i]*dy[l][k] / ( pow(dr[l][i] , 7) * pow(dr[l][k] , 7));
        G += -(25/4) * pow(b,10) * dz[l][i]*dz[l][k] / ( pow(dr[l][i] , 7) * pow(dr[l][k] , 7));
        }
      }
    }
  }
}
T[n] = 0.09094*(T[n]+G);
Tjf[n] = 0.09094*(T[n]/2);

E[n] = T[n]+V[n];

Esum += E[n];
Esum2 += pow(E[n],2);

}
double Nsamples = acc + rej;
printf("%f\n", Nsamples);
eff = acc / Nsamples;
printf("%f\n", eff);

//FILE *fp;
//fp=fopen("energies.csv","w+");
for (int n=0; n<Nsteps; n++){
  printf("%f, %f, %f \n", V[n],T[n],E[n]);
}

// We can now calculate the average value of the local energy and its error
double Evalue;
double deltaE;

Evalue = Esum/Nsteps;
deltaE = sqrt( 1/(Nsteps-1) * (Esum2/Nsteps - Evalue*Evalue) ); //for some reason it doesn't work
printf("value: %f error: %f \n", Evalue, deltaE);
printf("Esum: %f Esum2: %f \n", Esum, Esum2);

return 0;
}
